tools/convert_bigfile_gadget_format3.py

```diff
@@ -0,0 +1,271 @@
+                continue
+            ptype = int(ptype)
+            hname = names.get_hdf5_name(bname)
+            hdf5["PartType"+str(ptype)][hname] =  bf[block][startpart[ptype]:endpart[ptype]]
+        #Gadget-3 requires an InternalEnergy block for ICs, even if it is zero.
+        if "InternalEnergy" not in hdf5["PartType0"].keys():
+            hdf5["PartType0"]["InternalEnergy"] = np.zeros(endpart[0]-startpart[0])
+
+def compute_nfiles(npart):
+    """Work out how many files we need to split the snapshot into.
+       We want less than 2^31 bytes per data array, and we want a power of two,
+       so that we probably divide the particles evenly."""
+    nfiles = 1
+    #Largest possible data array: a 3-vector in double precision.
+    maxarray = np.max(npart) * 3 * 8
+    while maxarray // nfiles >= 2**31:
+        nfiles *=2
+    return nfiles
+
+def write_all_hdf_files(hdf5name, bfname):
+    """Work out which particle set goes to which HDF5 file and write it."""
+    bf = bigfile.BigFile(bfname, 'r')
+    nfiles = compute_nfiles(bf["Header"].attrs["TotNumPart"])
+    if not os.path.exists(hdf5name):
+        os.mkdir(hdf5name)
+    try:
+        mm = re.search("PART_([0-9]*)", bfname)
+        nsnap = '000'
+        if len(mm.groups()) > 0:
+            nsnap = mm.groups()[0]
+        hdf5name = os.path.join(hdf5name, "snap_"+nsnap)
+    except AttributeError:
+        hdf5name = os.path.join(hdf5name, bfname)
+    print("Writing %d hdf snapshot files to %s" % (nfiles, hdf5name))
+    for nn in range(nfiles):
+        write_hdf_file(bf, hdf5name, nn, nfiles)
+        print("Wrote file %d" % nn)
+
+def write_bigfile_header(hdf5, bf):
+    """Write out a header in the bigfile format. Default units are assumed."""
+    bf.create("Header")
+    battr = bf["Header"].attrs
+    hattr = hdf5["Header"].attrs
+    battr["BoxSize"] = hattr["BoxSize"]
+    #As a relic from Gadget-1, the total particle numbers
+    #are written as two separate 32 bit integers.
+    battr["TotNumPart"] = np.uint64(hattr["NumPart_Total_HighWord"])*2**32 + np.uint64(hattr["NumPart_Total"])
+    #Guess at the initial particle numbers
+    battr["TotNumPartInit"] = battr["TotNumPart"]
+    battr["TotNumPartInit"][0] += battr["TotNumPartInit"][4]
+    battr["TotNumPartInit"][4] = 0
+    battr["TotNumPartInit"][5] = 0
+    #Guess at this. It only really matters for the neutrino model, which isn't present in Gadget-3.
+    battr["TimeIC"] = np.min([hattr["Time"], 0.01])
+    try:
+        for attr in ["UnitLength_in_cm", "UnitMass_in_g", "UnitVelocity_in_cm_per_s"]:
+            battr[attr] = hattr[attr]
+    #Fall back to default unit system
+    except KeyError:
+        battr["UnitLength_in_cm"] = 3.085678e+21
+        battr["UnitMass_in_g"] = 1.989e43
+        battr["UnitVelocity_in_cm_per_s"] = 100000.
+    #Some flags
+    battr["UsePeculiarVelocity"] = 0
+    #Missing a factor of 1/Hubble, but doesn't matter.
+    battr["RSDFactor"] = 1./hattr["Time"]**2
```

Comment:
If we don't have the right value, just set a nan is better than a wrong value; at least it means if the value is used we will crash rather than give wrong answers siliently.

Reply:
Done!